pkg/apis/pipeline/v1beta1/taskrun_types.go

```diff
@@ -385,18 +385,35 @@ func (tr *TaskRun) IsCancelled() bool {
 
 // HasTimedOut returns true if the TaskRun runtime is beyond the allowed timeout
 func (tr *TaskRun) HasTimedOut(ctx context.Context) bool {
-	if tr.Status.StartTime.IsZero() {
+	startTime := tr.getStartTime()
+	if startTime.IsZero() {
 		return false
 	}
 	timeout := tr.GetTimeout(ctx)
 	// If timeout is set to 0 or defaulted to 0, there is no timeout.
 	if timeout == apisconfig.NoTimeoutDuration {
 		return false
 	}
-	runtime := time.Since(tr.Status.StartTime.Time)
+	runtime := time.Since(startTime.Time)
 	return runtime > timeout
 }
 
+func (tr *TaskRun) getStartTime() *metav1.Time {
+	startTime := tr.Status.StartTime
+	if startTime.IsZero() {
+		if len(tr.Status.RetriesStatus) == 0 {
+			return startTime
+		}
+		startTime = &metav1.Time{Time: time.Now()}
+	}
+	for _, retry := range tr.Status.RetriesStatus {
```

Comment:
I don't think the `TaskRun` controller should ever use any data from the `RetriesStatus`.

Retries cannot be defined in a `TaskRun`, they are a `Pipeline` feature really. 
Since this feature has been implemented in the `PipelineRun` controller as a `TaskRun` status reset + new `Pod`, the `RetriesStatus` was added to the `TaskRun` status to hold an history of past execution for the benefit of the `PipelineRun` controller as well as end users.

I don't know the motivation behind implementing retries like this - I assume there was a good reason for not creating new `TaskRuns` instead. But the fact remains that the `RetriesStatus` is written and managed by the `PipelineRun` controller, and I think they should be completely transparent from a `TaskRun` controller point of view.